### PR TITLE
add qcustomplot/2.1.0

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

closes https://github.com/conan-io/conan-center-index/issues/5983

Test package fails at runtime when qt is static (but not if shared, except on Linux). If someone knows how to make it work.

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Jihadist** [comment]
> Test package fails at runtime when qt is static (but not if shared, except on Linux). If someone knows how to make it work.

If you try to make gui app with conan-qt-static it will fail with the same problem

**Comment by ericLemanissier** [comment]
@Jihadist can you try to use `Q_IMPORT_PLUGIN (QXcbIntegrationPlugin)` in one of your cpp files ?

**Comment by Jihadist** [comment]
@ericLemanissier nope because generated cmake files don't provide direct path to plugin directory like this ```bin\archdatadir\plugins\styles``` so you cannot use Q_IMPORT_PLUGIN macro. I stucked on linkage errors with this macro.

**Comment by ericLemanissier** [comment]
can you please send the full log with`Q_IMPORT_PLUGIN (QXcbIntegrationPlugin)`, so that we have a chance to understand and fix the linkage errors?

**Comment by AndreyMlashkin** [comment]
@Jihadist, did it work for you for the old bincrafters recipe?

**Comment by SpaceIm** [comment]
> @ericLemanissier nope because generated cmake files don't provide direct path to plugin directory like this `bin\archdatadir\plugins\styles` so you cannot use Q_IMPORT_PLUGIN macro. I stucked on linkage errors with this macro.

I'm working on this, but it makes qt recipe very complex: https://github.com/SpaceIm/conan-center-index/blob/fix/qt6-components/recipes/qt/6.x.x/conanfile.py

**Comment by AndreyMlashkin** [comment]
@SpaceIm, there is no PR for it yet?
I am using a static build of qt, it is the last thing that is missing for me to start using conan-center version of qt recipe instead of bincrafters one

**Comment by ericLemanissier** [comment]
> Honestly, it would be easier to turn shared option of qt recipe to `True` by default

+10, I asked this for years. I acknowledge that it comes with its own set of problems because static dependencies are then embedded in the shared library, and they cannot be overridden without rebuilding qt, but that's less bad than the current situation

**Comment by Jihadist** [comment]
> 
> 
> I am getting an error
> 
> undefined reference to `qt_static_plugin_QXcbIntegrationPlugin()'
> 
> Should I add something to CMakeLists to make it work?

I mentioned it before https://github.com/conan-io/conan-center-index/pull/6029#issuecomment-874640381
Try to add target_link_directories with path something similar to path_to_conan_qt\bin\archdatadir\plugins

**Comment by Jihadist** [comment]
I found the solution but maybe qt5/6 recipe should be a bit expanded. You can found the QXcbIntegrationPlugin solution here https://github.com/Jihadist/StaticQtPlugin but it's dirty solution. 
Briefly, we need to link against
res/archdatadir/plugins/platforms/libqxcb.a and lib/libQt{$Qt_MAJOR_VERSION}XcbQpa.a and then use Q_IMPORT_PLUGIN (QXcbIntegrationPlugin)

**Comment by ericLemanissier** [comment]
@SpaceIm can you retrigger this PR, to see if the changes made in qt recipe this summer improved the situation ?

**Comment by SpaceIm** [comment]
Let's try again, I don't know whether qt recipe is fixed.

**Comment by ericLemanissier** [comment]
spoiler alert: it is not

**Comment by Jihadist** [comment]
nothing changes)

**Comment by SpaceIm** [comment]
Just to not close, still waiting a fix in qt recipe (but honestly change default shared value of qt recipe please, folks usually don't want qt static).

**Comment by ericLemanissier** [comment]
If it's the only blocking point, We could set qt as shared in this recipes configure ?

**Comment by SpaceIm** [comment]
Not 100% sure, but I'll try :)

**Comment by SpaceIm** [comment]
Seems to be an issue of qcustomplot since qt 6.2: https://www.qcustomplot.com/index.php/support/forum/2380

**Comment by SpaceIm** [comment]
Oh yes I remember now, it doesn't work on Linux with shared too.

**Comment by ericLemanissier** [comment]
Isn't it because the CI is headless ?
`qt.qpa.xcb: could not connect to display`

**Comment by SpaceIm** [comment]
> Isn't it because the CI is headless ?
> qt.qpa.xcb: could not connect to display

Why does it work on macOS & Windows?

**Comment by SpaceIm** [comment]
> ConanException: Error 3221225781 while executing bin\test_package
some missing dll in Debug build of qt?
https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/6029/5-configs/windows-visual_studio/qcustomplot/2.1.0//summary.json

**Comment by SpaceIm** [comment]
> > Isn't it because the CI is headless ?
> > qt.qpa.xcb: could not connect to display
> 
> Maybe, but why does it work in macOS & Windows agents?

I've tested on Ubuntu, and indeed tet package runs fine, so issue in CI likely due to headless.

**Comment by ericLemanissier** [comment]
yep, `glib-2.0-0.dll`

**Comment by SpaceIm** [comment]
glib is disabled in qt if not Linux/FreeBSD, but it's not disabled in harfbuzz. So I guess that https://github.com/conan-io/conan-center-index/pull/9739 indirectly broke current expectation in qt binaries.
But I still don't understand why there are errors in test package of qcustomplot for windows/debug only.

**Comment by ericLemanissier** [comment]
The existing shared build of qt have an old version of harfbuzz statically linked, which linked to shared glib/2.70.1 (on windows).
It only happens on windows because of https://github.com/conan-io/conan-center-index/blob/master/recipes/glib/all/conanfile.py#L63. Other platform always have a static glib. Maybe it happens only in debug because compiler/linker optimization put the loading of shared glib in a path which is not hit any more at run-time?

**Comment by SpaceIm** [comment]
There is no other solution than generating new packages of qt recipe.

**Comment by ericLemanissier** [comment]
https://github.com/conan-io/conan-center-index/pull/9782

**Comment by SSE4** [comment]
if you can upgrade to new `glib` version, it may allow to move things forward (latest `glib` releases support static builds on Windows)

**Comment by ericLemanissier** [comment]
the fact that default builds of windows glib are now static is the cause of the failure: every existing Windows binaries of all recipes (requiring glib) currently on conan-center are linking dynamically to glib, but the default glib is now static on windows, so basically all binaries requiring glib have to be regenerated, or there will be an error as soon as glib is [overridden to a version >= 2.71.1](https://github.com/conan-io/conan-center-index/blob/master/recipes/glib/all/conanfile.py#L65)

**Comment by ericLemanissier** [comment]
qt 6 binaries have been fixed with https://github.com/conan-io/conan-center-index/pull/9782. Can we retry ?

**Comment by SSE4** [comment]
it seems like package_id model is either too flawed, or too strict, and it's hard to find the balance
as I understand, what happens, more or less, is:
- we compile `glib:shared=True`
- compile `Qt` against `glib:shared=True`
- at this point, glib's option `glib:shared=True` is erased from the Qt's package id, unless it uses something like `full_package_mode`
- compile `QCustomPlot` overriding `glib:shared=False`
- a final binary of `QCustomPlot` still require `glib2.dll`, but it's not there, as conan thinks `glib` is static
- but actually, compiled `Qt` library references `glib2.dll`, because it was compiled against shared `glib` before
- there is no way to distinguish these things, as `Qt` package ID is the same regardless of `glib:shared` option

might be also worth checking https://github.com/conan-io/conan/issues/9712
TLDR: `Qt` might need to add `self.info.requires["glib"].full_package_mode()` to its `package_id`

**Comment by ericLemanissier** [comment]
I agree with all your message, except your last sentence `Qt might need to add self.info.requires["glib"].full_package_mode() to its package_id`
The issue has nothing to do with glib or qt recipe. It showed up on this PR because there was a change in the default value of glib:shared option, but it can happen any time a user changes the :shared value of any package (A) in the dependency tree of any recipe, if said package (A) is a requirement of another package B which has B:shared=True

**Comment by SpaceIm** [comment]
yes, this why https://github.com/conan-io/conan-center-index/issues/8987 would be a solution

**Comment by SSE4** [comment]
yes, it's a generic problem, that couldn't be fixed in conan 1.x unfortunately, and it affects all the recipes (with at least one dependency with `shared` option). basically, when you specify `xxx:shared=True/False` you don't get it applied to dependencies, because it was erased from the `package_id`. it's terrible, but nothing to do except rebuilding `Qt` and other packages once it happens...

**Review by ericLemanissier — APPROVED** [review]

**Review by AndreyMlashkin — APPROVED** [review]

**Review by Jihadist — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
